Feature/security headers

DanielHudson2 opened the pull request:
Added an action for `wp_headers` to set security headers

For now have only added a basic default `'X-Frame-Options' = 'SAMEORIGIN'` to prevent click jacking @edjeavons could do with some input on others from a server side perspective

This came from checking the security headers on the eighteen73 site https://securityheaders.com/?q=https%3A%2F%2Feighteen73.co.uk

Also added a filter so this can be overridden on a per site basis if needed

DanielHudson2 committed:
New: added class for adding security headers

DanielHudson2 committed:
New: document security headers

brettsmason reviewed (commented):
Looks good to me. My only comment is should we have a filter to enable/disable the same as others, eg `orbit_enable_security_headers`. I'll leave Ed to comment on what the defaults should be.

edjeavons committed:
Additional default security headers

edjeavons commented:
This is a really good idea, thanks @DanielHudson2 

I've added some additional suggestions for the default header set. We'll need to give `Content-Security-Policy` a little thought to consider what's safe because I've definitely had problems trying to jump directly to generalised rules with that before.

I've added a `is_ssl()` check around the `Strict-Transport-Security` header to ensure it doesn't harm websites before devs are ready for it, but you and @brettsmason might a preferred way to test for whether or not requests are using SSL.

brettsmason commented:
@edjeavons Could we have a catch up on this at some point to try and get this through?

edjeavons committed:
Set Cache-Control by default

This is expected for PCI tests

edjeavons committed:
Typo correction in Headers.php

edjeavons committed:
Use a more permissive CSP by default

This can be considered a placeholder for websites to customise based on their own needs.

edjeavons committed:
Merge branch 'develop' into feature/security-headers

edjeavons committed:
Package updates

edjeavons committed:
Remove forced Cache-Control header

We need to prove this won’t affect full page caching before adding it

edjeavons committed:
Correction to using the wp_headers filter

edjeavons committed:
Improve the default CSP

edjeavons committed:
Add CSP rule necessary for functionality

edjeavons reviewed (approved):
I've refined a couple of things and tested it on a website. It's OK to merge as soon as you're ready.
